kv/kvserver: TestFlowControlRaftSnapshotV2 failed

kv/kvserver.TestFlowControlRaftSnapshotV2 [failed](https://tanzanite.cluster.engflow.com/invocations/default/541a96db-f870-4480-9ef0-f2678ac0a3ea?testReportRun=1&testReportShard=7&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on release-25.1.3-rc @ [abdedd3dce8c0734e8b57af59e2b0262b0a8e82b](https://github.com/cockroachdb/cockroach/commits/abdedd3dce8c0734e8b57af59e2b0262b0a8e82b):

Fatal error:

```
panic: test timed out after 40m0s
running tests:
	TestFlowControlRaftSnapshotV2 (39m34s)
	TestFlowControlRaftSnapshotV2/v2_enabled_when_leader_level=1 (39m34s)
	TestFlowControlRaftSnapshotV2/v2_enabled_when_leader_level=1/kvadmission.flow_control.mode=apply_to_all (39m23s)

```
Stack: 

```
goroutine 89506595 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
* 
* goroutine 4895515 [chan receive, 2 minutes]:
* github.com/cockroachdb/pebble.(*fileCacheShard).releaseLoop(0xc01ae59d90)
* 	external/com_github_cockroachdb_pebble/file_cache.go:399 +0x97
* created by github.com/cockroachdb/pebble.(*fileCacheShard).init in goroutine 4176005
* 	external/com_github_cockroachdb_pebble/file_cache.go:390 +0xe5
* 
* goroutine 5108539 [chan receive, 2 minutes]:
* github.com/cockroachdb/pebble.(*fileCacheShard).releaseLoop(0xc01bfd4e70)
* 	external/com_github_cockroachdb_pebble/file_cache.go:399 +0x97
* created by github.com/cockroachdb/pebble.(*fileCacheShard).init in goroutine 4176005
* 	external/com_github_cockroachdb_pebble/file_cache.go:390 +0xe5
* 
* goroutine 5109926 [select]:
* github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue.func2()
* 	pkg/util/admission/work_queue.go:424 +0x85
* created by github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue in goroutine 4176005
* 	pkg/util/admission/work_queue.go:421 +0x3be
* 
* goroutine 5034717 [select]:
* github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue.func2()
* 	pkg/util/admission/work_queue.go:424 +0x85
* created by github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue in goroutine 4176005
* 	pkg/util/admission/work_queue.go:421 +0x3be
* 
* goroutine 4808436 [select, 2 minutes]:
* database/sql.(*DB).connectionOpener(0xc003f3b110, {0x902fc20, 0xc00beb6820})
* 	GOROOT/src/database/sql/sql.go:1246 +0x87
* created by database/sql.OpenDB in goroutine 4176005
* 	GOROOT/src/database/sql/sql.go:824 +0x14c
* 
* goroutine 5108536 [chan receive, 2 minutes]:
* github.com/cockroachdb/pebble.(*fileCacheShard).releaseLoop(0xc01bfd4c60)
* 	external/com_github_cockroachdb_pebble/file_cache.go:399 +0x97
* created by github.com/cockroachdb/pebble.(*fileCacheShard).init in goroutine 4176005
* 	external/com_github_cockroachdb_pebble/file_cache.go:390 +0xe5
* 
* goroutine 5109753 [select]:
* github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue.func2()
* 	pkg/util/admission/work_queue.go:424 +0x85
* created by github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue in goroutine 4176005
* 	pkg/util/admission/work_queue.go:421 +0x3be
* 
* goroutine 5112924 [select]:
* github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker.func1()
* 	external/com_github_cockroachdb_pebble/vfs/disk_health.go:254 +0xbc
* created by github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker in goroutine 4176005
* 	external/com_github_cockroachdb_pebble/vfs/disk_health.go:249 +0x58
* 
*
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=1</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139229 kv/kvserver: TestFlowControlRaftSnapshotV2 failed [C-bug C-test-failure O-robot P-2 T-kv X-duplicate branch-release-25.1 target-release-25.1.1 target-release-25.1.2 v25.2.0-prerelease]
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestFlowControlRaftSnapshotV2.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-49106

--- a/pkg/sql/distsql/server.go
+++ b/pkg/sql/distsql/server.go
@@ -624,6 +624,10 @@ func (ds *ServerImpl) SetupFlow(
 	// Note: the passed context will be canceled when this RPC completes, so we
 	// can't associate it with the flow since it outlives the RPC.
 	ctx = ds.AnnotateCtx(context.Background())
+	// Ensure that the flow respects the node being shut down. Note that since
+	// the flow outlives the RPC, we cannot defer the cancel function, so we
+	// simply ignore it.
+	ctx, _ = ds.Stopper.WithCancelOnQuiesce(ctx)
 	if err := func() error {
 		// Reserve some memory for this remote flow which is a poor man's
 		// admission control based on the RAM usage.

